### PR TITLE
SessionMiddleware and storing session in request context

### PR DESCRIPTION
**Pull request by adam-fowler**
- Add new request context protocol `SessionRequestContext` which requires a `SessionContext` type which holds a session data. The session data type is a generic parameter of these two. 
- Add middleware `SessionMiddleware` which loads the session data from storage into the context and then after processing the request if it has been edited it will save it back into the storage. If this is a new session then it will return a set-cookie header.

Read only access to session data can be done with 
```swift
context.sessions.session
```
Write access to session data is available with 
```swift
context.sessions.setSession(_:)
```
If you have a more complex editing operation you can use
```swift
context.sessions.withLockedSession { session in
    editSession(session)
}
```
I also edited the `SessionAuthenticator` middleware to work with the new context

### CODE REVIEW

**Comment by Joannis** [comment]
Looks great already :)

File: Sources/HummingbirdAuth/Sessions/SessionContext.swift [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        }
+    }
+}
+
+/// Session context
+///
+/// Holds reference to session data protected by a lock to avoid concurrent access
+public struct SessionContext<Session: Sendable & Codable>: Sendable {
+    @usableFromInline
+    let _storage: NIOLockedValueBox<SessionData<Session>?>
```
[the remark this answers is not in the document]
Comment (adam-fowler) [reply Sources/HummingbirdAuth/Sessions/SessionContext.swift:68]: We don't have an EventLoop to bind to

File: Sources/HummingbirdAuth/Sessions/SessionContext.swift [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+/// Session data
+@dynamicMemberLookup
+public struct SessionData<Session: Sendable & Codable>: Codable, Sendable {
+    @usableFromInline
+    var object: Session
+    @usableFromInline
+    var edited: Bool
+    /// When session will expire
+    public var expiresIn: Duration?
```
[the remark this answers is not in the document]
Comment (adam-fowler) [reply Sources/HummingbirdAuth/Sessions/SessionContext.swift:27]: Should the expires value be based on when this object is created or when the request returns? I'm not sure.

File: Sources/HummingbirdAuth/Sessions/SessionContext.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (Joannis) [reply Sources/HummingbirdAuth/Sessions/SessionContext.swift:68]: Right..

File: Sources/HummingbirdAuth/Sessions/SessionContext.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (Joannis) [reply Sources/HummingbirdAuth/Sessions/SessionContext.swift:27]: That's a good question!

[the remark this answers is not in the document]
Comment (adam-fowler) [reply Sources/HummingbirdAuth/Sessions/SessionContext.swift:27]: In most cases I don't think it matters. But I would probably go for when the response is returned which is what happens now.

**Review by Joannis — APPROVED** [review]

**Pull request merged**